From d2f0e0615417ae62ae46586d9413af9c736cb038 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jochen=20Sch=C3=A4fer?= <josch1710@live.de>
Date: Sun, 12 Apr 2026 09:40:47 +0200
Subject: [PATCH] Macos: Moved platform specific headers to header section

---
 src/emx/src/libcrossemx/crossemx.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/emx/src/libcrossemx/crossemx.c b/src/emx/src/libcrossemx/crossemx.c
index e00c21a..edf762a 100644
--- a/src/emx/src/libcrossemx/crossemx.c
+++ b/src/emx/src/libcrossemx/crossemx.c
@@ -31,6 +31,10 @@ Boston, MA 02111-1307, USA.  */
 #include <sys/wait.h>
 #include <errno.h>
 #include <libiberty.h>
+#if defined __APPLE__
+#include <mach-o/dyld.h>
+#include <libgen.h>
+#endif
 
 #include "crossemx.h"
 
@@ -273,9 +277,6 @@ int _execname (char *buf, size_t bufsize)
   return 0;
 }
 #else
-#include <mach-o/dyld.h>
-#include <libgen.h>
-
 int _execname (char *buf, size_t bufsize)
 {
   char path[1024];
-- 
2.50.1 (Apple Git-155)

